src/emit/mod.rs

```diff
@@ -5505,6 +5505,200 @@ pub trait TargetRuntime<'a> {
         vector.into()
     }
 
+    fn signed_ovf_detect(
+        &self,
+        bin: &Binary<'a>,
+        l: PointerValue<'a>,
+        r: PointerValue<'a>,
+        o: PointerValue<'a>,
+        mul_bits: u32,
+        left: IntValue<'a>,
+        right: IntValue<'a>,
+        bits: u32,
+        function: FunctionValue,
+    ) -> IntValue<'a> {
+        // In llvm, Two's complement representation is used.
+        // We check for overflow based on the facts:
+        //  - * - = +
+        //  + * + = +
+        //  - * + = - (if op1 or op2 != 0)
+        // if one of the operands is zero, discard the last rule.
+        bin.builder.build_call(
+            bin.module.get_function("__mul32").unwrap(),
+            &[
+                bin.builder
+                    .build_pointer_cast(
+                        l,
+                        bin.context.i32_type().ptr_type(AddressSpace::Generic),
+                        "left",
+                    )
+                    .into(),
+                bin.builder
+                    .build_pointer_cast(
+                        r,
+                        bin.context.i32_type().ptr_type(AddressSpace::Generic),
+                        "right",
+                    )
+                    .into(),
+                bin.builder
+                    .build_pointer_cast(
+                        o,
+                        bin.context.i32_type().ptr_type(AddressSpace::Generic),
+                        "output",
+                    )
+                    .into(),
+                bin.context
+                    .i32_type()
+                    .const_int(mul_bits as u64 / 32, false)
+                    .into(),
+            ],
+            "",
+        );
+
+        let res = bin.builder.build_load(o, "mul");
+
+        let error_block = bin.context.append_basic_block(function, "error");
+        let return_block = bin.context.append_basic_block(function, "return_block");
+
+        // Extract sign bits of the operands and the result
+        let left_sign_bit = self.extract_sign_bit(bin, left, left.get_type());
+        let right_sign_bit = self.extract_sign_bit(bin, right, right.get_type());
+        let res_sign_bit = if mul_bits == bits {
+            // If no extension took place, get the MSB.
```

Comment:
You are not extracting the most significant bit. You are extracting the sign bit. If you have `int32 g = 4`, the most significant bit is the third, while the sign bit is the 32nd.